R/paginate.R

```diff
@@ -0,0 +1,336 @@
+    ))
+  }
+
+  next_request <- req$policies$paginate$next_request
+  next_request(resp = resp, req = req)
+}
+
+#' @rdname paginate_perform
+paginate_n_pages <- function(resp, req, max_pages) {
+  check_response(resp)
+  check_request(req)
+
+  page_size <- req$policies$paginate$page_size
+  total <- req$policies$paginate$total
+
+  if (is.null(total) || is.null(page_size)) {
+    return(max_pages)
+  }
+
+  body_parsed <- resp_body_json(resp)
+  total <- purrr::pluck(body_parsed, total)
+  if (is.null(total)) {
+    return(max_pages)
+  }
+  n_pages <- ceiling(total / page_size)
+
+  min(n_pages, max_pages)
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_url <- function(req,
+                                  next_url,
+                                  ...,
+                                  page_size = NULL,
+                                  total = NULL) {
+  check_character(next_url)
+
+  next_request <- function(req, resp) {
+    body_parsed <- resp_body_json(resp)
+    next_url <- purrr::pluck(body_parsed, next_url)
+
+    if (is.null(next_url)) {
+      return(NULL)
+    }
+
+    req_url(req, next_url)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_offset <- function(req,
+                                offset,
+                                page_size,
+                                total = NULL) {
+  check_param(offset, value = FALSE, allow_null_value = TRUE)
+
+  cur_offset <- 0L
+  env <- current_env()
+  next_request <- function(resp, req) {
+    cur_offset <- get("cur_offset", envir = env)
+    new_offset <- cur_offset + page_size$value
+    assign("cur_offset", new_offset, envir = env)
+
+    req_set_param(req, offset, new_offset)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_token <- function(req,
+                                    token_field,
+                                    next_token_field,
+                                    page_size = NULL,
+                                    total = NULL) {
+  check_param(token_field, value = FALSE, allow_null_value = TRUE)
+  check_character(next_token_field)
+
+  next_request <- function(req, resp) {
+    body_parsed <- resp_body_json(resp)
+    next_token <- purrr::pluck(body_parsed, next_token_field)
+
+    if (is.null(next_token)) {
+      return(NULL)
+    }
+
+    req_set_param(req, token_field, next_token)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' Specify a request parameter
+#'
+#' @param name Name of the query resp. header parameter.
+#' @param path Path of the parameter in the body.
+#' @param value The value of the parameter.
+#' @param error_call
+#'
+#' @return A parameter object.
+#' @export
+#'
+#' @examples
+#' in_query("start", 20)
```

Comment:
I wonder if we need these helpers. It's not too awful to supply an anonymous function if there's an existing response helper:

```R
page_size = \(resp) resp_url_query(resp, "limit", 150L),
next_page = \(resp) resp_link_url(resp, "next")
```

Then we'd just need something extra in `resp_body_json()` that let you drill down to a specific component.

Reply:
(That makes me wonder if `resp_link_url()` should actually be `resp_header_link()`)